### DISCUSSION TITLE
How to extract `react-dom-server.browser` and remove `react-dom-server-legacy.browser` from client bundle size?

### DISCUSSION THREAD

**Discussion by bolbovic** [vercel/next.js#49795] [Help]
### Summary

I am using some LiquidJS to render some interactive text and I need to use the function `renderToString` (can't use `createRoot` here sadly).

Before, I upgraded to React18 and Next13, I had already the `react-dom-server.browser.production.js` which is not great but I accepted as it was. But with the upgrade, I have now both `react-dom-server.browser` and `react-dom-server-legacy.browser` even if I have `experimental: { legacyBrowsers: false },` in next config.
The bundle also got bigger so I have about 20kb more in the _app package and it starts to sting a bit.
So first question would be how do I get rid of `react-dom-server-legacy` ?

Second problem would be to remove the `react-dom-server.browser.production.js` from the app framework chunk and have it only for the few pages I am using it. How can I achieve this? I tried to load the the function dynamically like:
```
async function renderToString(element: React.ReactElement) {
  const s = (await import("react-dom/server")).default;
  return s.renderToString(element);
}
```
Tried a few variation of it (specific file for example) but the `react-dom-server` would always be in the framework chunk.
Obviously, if I remove the code for `renderToString` at this specific place in the code, the bundle doesn't include neither `react-dom-server` or `react-dom-server-legacy` so the problem is really there and not anywhere else in the app.

Any idea?

Thanks a lot in advance!

**Comment by zzzzzzxXxzzzzzz** [comment]
> ### Summary
> 
> I am using some LiquidJS to render some interactive text and I need to use the function `renderToString` (can't use `createRoot` here sadly).
> 
> Before, I upgraded to React18 and Next13, I had already the `react-dom-server.browser.production.js` which is not great but I accepted as it was. But with the upgrade, I have now both `react-dom-server.browser` and `react-dom-server-legacy.browser` even if I have `experimental: { legacyBrowsers: false },` in next config.
> The bundle also got bigger so I have about 20kb more in the _app package and it starts to sting a bit.
> So first question would be how do I get rid of `react-dom-server-legacy` ?
> 
> Second problem would be to remove the `react-dom-server.browser.production.js` from the app framework chunk and have it only for the few pages I am using it. How can I achieve this? I tried to load the the function dynamically like:
> ```
> async function renderToString(element: React.ReactElement) {
>   const s = (await import("react-dom/server")).default;
>   return s.renderToString(element);
> }
> ```
> Tried a few variation of it (specific file for example) but the `react-dom-server` would always be in the framework chunk.
> Obviously, if I remove the code for `renderToString` at this specific place in the code, the bundle doesn't include neither `react-dom-server` or `react-dom-server-legacy` so the problem is really there and not anywhere else in the app.
> 
> Any idea?
> 
> Thanks a lot in advance!
>

**Comment by oleksandr-p-wl** [comment]
Hi @bolbovic ! Any luck with this?

**Comment by pavel-blagodov** [comment]
Exclude it explicitly via your build tool
`
        external: ['react-dom/server']`

**Comment by KATT** [comment]
Heya. Late to the party here. Anyone ever solved this?
